### DISCUSSION TITLE
Architectural Drivers

### DISCUSSION THREAD

**Discussion by changjurhee** [changjurhee/AONE#2] [General]
- Are the QA requirements “actionable”? In other words, are they expressed in such a way that the team will be able to determine if a given design supports these drivers or not?
- Do the drivers seem to relate to the overall objectives of the project?
- Are the measures clearly derived from the overall goals of the project?
- Are the functional requirements understood?
- Is there a mechanism for prioritizing the requirements?

**Comment by BrunoYoo** [comment]
![image](https://github.com/changjurhee/AONE/assets/101427090/5bf95878-85e3-41f8-a27b-eb9fd0978008)

<html>
<body>
<google-sheets-html-origin>

Type | Category | Description
-- | -- | --
FR | User account / Register, update, change | System shall provide to register/ store user account with Email and password
FR | User account / Register, update, change | System shall provide to update / change user account for Email and password
FR | User account / Register, update, change | System shall provide to add / update user information such as Name, First Name, Address
FR | User account / Register, update, change | System shall provide the ability for the administrator to view user accounts and disable, enable, and delete user accounts.
FR | User account / Contact identification | System shall assign the user a unique contact identification name (contact identifier) for each user account using e-mail address or some other name chosen by the user if it does not conflict with other user’s contact identifiers already in the system.
FR | User account / IP Directory service | System shall provide a directory services capability that maps a device’s contact identifier to its current IP address.
FR | User account / IP Directory service | System shall notify the directory services component of the backend server of IP changes.
FR | User account / Contact lists | System shall provide to create contact list by searching users by name, first name, address, email or contact identifier
FR | User account / Contact lists | System shall provide to show contact list that user created
FR | General application | Application shall be brought to the foreground during an incoming call.
FR | Connections | System shall allow user to input server IP address for interactions.
FR | Video | System shall capture video from Laptop's camera for direct call & video conference
FR | Video | System shall show direct call / video conference participants streaming video clip transferred via WiFi Network
FR | Audio | System shall capture audio from Laptop's microphone for direct call & video conference
FR | Audio | System shall play direct call / video conference audio transferred via WiFi Network
FR | Audio | System shall provide acoustic echo cancellation (AEC) while captureing audio
FR | Audio | System shall provide voice activation detection (VAD) while capturing audio and the user application shall not send voice data during periods of silence when the user is not speaking.
FR | Direct call | System shall provide user to request Direct call with one participant contact identifier
FR | Direct call | System shall provide user to request Direct call using using contacts list
FR | Direct call | System shall show user of Direct call status and outcome (answered, busy or rejected) during the call initiation
FR | Direct call | System shall notify user of Direct call status (Termination, Lost connection, conference status) during the active call status
FR | Direct call | System shall show Direct call participant with their contact data with Name instead of contact identifier
FR | Direct call | System shall notify users of lost Direct call connection
FR | Direct call | System shall provide user to end the Direct call at any time.
FR | Conference call | System shall provide to request/schedule Conference call with participant contact identifiers for particular date, time, and duration.
FR | Conference call | System shall manage conference call IDs and distribute to participant via user application in an upcoming conference meeting list or via email
FR | Conference call | System shall show Conference call participants info with their contact data with Name instead of contact identifier
FR | Conference call | System shall notify user of Conference call status (Connection, Termination, Lost connection, conference status) with UX (Graphic / Text Notification) to user
FR | Conference call | System shall provide user to end the Conference call at any time.
FR | Conference call | System shall provides a graphical highlight to identify which participant is currently speaking.


</body>
</html>

<html>
<body>
<google-sheets-html-origin>

Type | Type | Description | QA Scenario
-- | -- | -- | --
QA | Security | System shall handle the user’s password in secure way | When the server's user contact list is leaked through an invalid path from outside the system, the origin of the user's password must not be deduced.
QA | Security | System shall transfer data for sessions, video data, audio streaming data between server & client in secure way. | When video and audio data between clients or session data between server and client are leaked through an invalid path from outside the system, the original contents must not be estimated.
QA | Performance | Provide reliable signaling for both inbound and outbound calls including dropped call detection. | When a user makes or terminates a direct call using the system in the runtime environment, the frequency of accurately responding to the call / end call attempt must be 99.9% or higher.
QA | Performance | Provide high quality low latency with strategies for dealing with packet or data loss and jitter. | When a user makes a video call using the system in a runtime environment, real-time video latency should be less than 1 second.
QA | Availability | Support a video call if there is enough available bandwidth. | When a user makes a direct call or conference call using the system in a runtime environment, an error must be notified to the user within 500 ms in a state where not enough bandwidth.
QA | Performance | Adjust bandwidth use as necessary to maintain communication quality. | When a user makes a direct call or conference call using the system in a network transmission speed environment of 100 kbps or less, real-time video latency of less than 1 second must be guaranteed even when sufficient bandwidth is not secured.
QA | Modifiability | Provide the ability for the developer/maintainer to easily add or replace codecs | It should be possible within 5 hours for the system developer to change the codec for video transmission after the initial development.
QA | Modifiability | Provide the ability for the developer/maintainer to easily add or replace encryption algorithms. | It should be possible within 5 hours for the system developer to change the encription algorithm for video transmission after the initial development.


</body>
</html>

<html>
<body>
<google-sheets-html-origin>

Type | Description
-- | --
Constraint | NOT allowed to use any Session Initiation Protocols (SIP) or any VoIP libraries in source code or library form
Constraint | Operating system should be Microsoft Windows 10 or 11.
Constraint | Developer tool should be Microsoft Visual Studio Community 2022 (64-bit) Version 17.5.5 or later
Constraint | Utilize LG’s Wi-Fi network for normal application development, testing and demonstrations.  If desired, LG may provide teams with individual Wi-Fi routers.
Constraint | Use their own laptops for software development and to execute the user application and server software.
Constraint | Video is captured using Laptop's camera
Constraint | Audio is captured using Laptop's microphone
Constraint | The system must be completed and verified before the end of the training course.


</body>
</html>

**Reply by ch0115** [reply]
Design Decision
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 12">
<link id=Main-File rel=Main-File
href="file:///C:\Users\chkim\AppData\Local\Temp\msohtmlclip1\01\clip.htm">
<link rel=File-List
href="file:///C:\Users\chkim\AppData\Local\Temp\msohtmlclip1\01\clip_filelist.xml">



</head>

<body link=blue vlink=purple>



Priority | Functional Requirement | Design Decision
-- | -- | --
? | Application | Structural design according to foreground and background operation



</body>

</html>

**Comment by changjurhee** [comment]
- Reference :
  - LG Arch 2023 Project Description.pdf
  - LG May 2023 Lecture Architecture Project Intro V1.0.pdf
  - LG Architecture Project Introduction.mp4

## Function Requirements
|No|Function Requirement|Descriptions|Link|
|--|--|--|--|
|1|User registration and contact identifier assignment. | Register each user’s e-mail address and password. The system shall ensure that the user’s password is secure.||
|2|Change user information|The ability for the user to update their e-mail address and password. ||
|3|Recover password|The ability for the user to change their password or reset it in the event it is lost. ||
|4|After successful registration the system shall assign the user a unique contact identification name (contact identifier) this can be the users e-mail address or some other name chosen by the user if it does not conflict with other user’s contact identifiers already in the system. |||
|5|The ability to request a conference call identifier (like a contact identifier) for a particular date, time and duration from the server that is then distributed to participants via user application in an upcoming conference meeting list or via e-mail. |||
|6|The ability to create a contacts list. |To add a contact to the list the user can search users by last name, first name, address, e-mail, or contact identifier. When a contact is associated with a contact identifier the VoIP application shall display the contact’s name instead of the contact identifier.||
|7|Receive call with application minimized|||
|8|The ability to initiate a call using a contact identifier or the contacts list.  |During the call initiation, the user shall be presented with call status and outcome (answered, busy or rejected). During call initiation the user shall have the ability to end the call at any time.||
|9|Provide the ability to accept or reject calls while not in a call. | Application shall show the caller’s contact identifier or contact name during an incoming call.|| 
|10|Adding and modifying user contacts|||
|11|Initiate call using contact identifier and contacts list|||
|12|The server shall provide a directory services capability that maps a device’s contact identifier to its current IP address.|Server||
|13|Ability to schedule and support video conference calls. Conference calls shall be scheduled with participants via a conference identifier. |Server||
|14|The ability for the administrator to view user accounts and disable, enable, and delete user accounts. |Server||
|15|Application shall be brought to the foreground during an incoming call. |||
|16|Notify the directory services component of the backend server of IP changes. |||
|17|Provide the ability to terminate a call at any time while in a call. If a call is terminated by one user, the other caller shall be notified.|||
|18|Provide acoustic echo cancellation (AEC). |||
|19|Provide voice activation detection (VAD). |The user application shall not send voice data during periods of silence when the user is not speaking. ||
|20|Notify users of lost connections.|||

## Quality Attributes
|No|Function Requirement|Descriptions|Link|
|--|--|--|--|
|1|performance|Provide high quality low latency with strategies for dealing with packet or data loss and jitter.||
|2|performance|Adjust bandwidth use as necessary to maintain communication quality. |||
|3|performance|Provide reliable signaling for both inbound and outbound calls including dropped call detection. ||
|4|security|Provide secure communication. ||
|5|security|The system shall ensure that the user’s password is secure.||
|6|capacity|Support a video call if there is enough available bandwidth.||
|7|Modifiability|[Server Application] Provide the ability for the developer/maintainer to easily add or replace codecs and/or encryption algorithms.||
|8|Modifiability|[Window Application] Provide the ability for the developer/maintainer to easily add or replace codecs and/or encryption algorithms.||

## Design Constraint
|No|Function Requirement|Descriptions|Link|
|--|--|--|--|
|1|The system shall run on Microsoft Windows 10 or 11 platforms.|||
|2|The system shall be implemented using Visual C++.|||

## Design Decision
|Priority|Quality Attributes|Design Decision|Link|
|--|--|--|--|
| 1 |  performance | apply MESH architecture | |

**Comment by junsoopark** [comment]
# Project Overview
robust and secure VoIP and video communication system
robust
secure

Microsoft Windows application and a backend server that provides various capabilities.
ms user application
backend server

SecureMeet Inc. will select the best architecture/solution
best architectire

Ensure quality voice and video communication that is also secure and reliable.
quality voice and video
reliable

Ensure that the system backend server and user application minimizes network bandwidth as necessary to maintain communication between the parties. 
minimize network bancwidth
maintain comm

Provides the ability to easily add or replace codecs and/or encryption algorithms.
replace codec
reploace encryption algorithms

user registration, contact identification assignment, directory services, and functionality to support video conference calls.
user registration
contact identification assignment
directory services
support video conference calls

During a video call, the video of the person speaking should be displayed to all participants. 
person speaking should be displayed to all

All teams will interact, communicate their client related questions or concerns with a single SecureMeet point of contact.
daniel?

Please note that teams are NOT allowed to use any Session Initiation Protocols (SIP) or any VoIP libraries in source code or library form.
Session Initiation
VoIP libraries

# System Description

## Microsoft Windows User Application

1. The ability for the user to register with the system. 

2. 
The ability to request a conference call identifier (like a contact identifier)
that is then distributed to participants via user application in an upcoming conference meeting list or via e-mail.
an upcoming conference meeting list

3.
The ability to create a contacts list.
user can search users
When a contact is associated with a contact identifier the VoIP application shall display the contact’s name instead of the contact identifier.
a contacts list
a contact
a contact identifier

4. All communications shall:
a. high quality low latency with strategies for dealing with packet or data loss and jitter.
b. Support a video call if there is enough available bandwidth.
c. Adjust bandwidth use as necessary to maintain communication quality.
d. Provide secure communication.
e. Provide reliable signaling for both inbound and outbound calls including dropped call detection.
f. Notify users of lost connections.
g. rovide acoustic echo cancellation (AEC).
h. Provide voice activation detection (VAD).  The user application shall not send voice data during periods of silence when the user is not speaking.

5. initiate a call using a contact identifier or the contacts list.
a call
a contact identifier
the contacts list

the user shall be presented with call status and outcome (answered, busy or rejected). 
call status
outcome (answered, busy or rejected)

During call initiation the user shall have the ability to end the call at any time.
end the call at any time.

6. Provide the ability to accept or reject calls while not in a call. Application shall show the caller’s contact identifier or contact name during an incoming call.
while not in a call.
the ability to accept or reject calls 
show the caller’s contact identifier or contact name during an incoming call.

7. Provide the ability to terminate a call at any time while in a call. If a call is terminated by one user, the other caller shall be notified.
while in a call
the ability to terminate a call
the other caller shall be notified.

8. Notify the directory services component of the backend server of IP changes.
IP changes.

9. Provide the ability for the developer/maintainer to easily add or replace codecs and/or encryption algorithms.
replace codecs and/or encryption algorithms.

10. Application shall be brought to the foreground during an incoming call.
brought to the foreground
during an incoming call.

## Server Application

1. Register each user’s e-mail address and password. The system shall ensure that the user’s password is secure.
reg e-mail address and password.
password is secure.

2. The server shall provide a directory services capability that maps a device’s contact identifier to its current IP address.
a directory services
a device’s contact identifier to its current IP address.

3. Ability to schedule and support video conference calls. Conference calls shall be scheduled with participants via a conference identifier.
schedule and support video conference calls.
with participants via a conference identifier.


4. The ability for the administrator to view user accounts and disable, enable, and delete user accounts.
the administrator
view user accounts and disable, enable, and delete user accounts.

5. Provide the ability for the developer/maintainer to easily add or replace codecs and/or encryption algorithms.
the developer/maintainer
easily add or replace codecs and/or encryption algorithms.

**Comment by du0san** [comment]
FR-01 : [User register] User application provides a function for the user to register in the system.
FR-02: [User's information] The user application requires the user's email and password during registration, while also allowing them to optionally provide their last name, first name, and address.
FR-03: [Password's encryption] The system shall ensure that the user’s password is secure.
FR-04 : [Updating user's information] User application provides a function for the user to update their e-mail address and password
FR-05 : [change password] User application provides a function for the user to change their password or reset it.
FR-06 : [Reset password] The user application provides a function for the user to reset their password after the user confirms their email.
FR-07 : [unique contact identifier] After a successful registration, the system will assign the user a unique contact identifier name, such as their email address or a user-selected name, provided it does not conflict with other existing contact identifiers.
FR-08 : [Registing conference call] The user application provide the function for user to request a conference call identifier for a particular date, time and duration.
FR-09 : [Notification conference call] The user application notify an upcoming conference meeting list of user and send e-mail.
FR-10 : [Creating contacts list] : The user application create a contacts list. To add a contact to the list the user can search users by last name, first name, address, e-mail, or contact identifier.
FR-11 : [Contact's expression] When a contact is associated with a contact identifier the user application shall display the contact’s name instead of the contact identifier.
FR-12 : [Providing AEC] The user application provide acoustic echo cancellation (AEC) in communication.
FR-13 : [Providing VAD] The user application provide voice activation detection(VAD) in communication.
FR-14 : [Communication's encryption] The user application provide secure communication
FR-15 : [Call's initiation] The user application initiate a call using a contact identifier or the contacts list.
FR-16 : [Calls' status] The user application display  call status and outcome (answered, busy or rejected) to user during call initiation.
FR-17 : [End of call's initiation] The user application provide the function for user to end the call at any time during call initiation.

**Comment by codekistone** [comment]
FR
Category | Description
-- | --
Client: | User registration feature.
  | Password security verification feature.
  | Email address and password update feature.
  | Password change and reset feature.
  | Allocation of a unique contact identifier to users.
  | Requesting a conference call identifier for a specific date, time, and duration.
  | Contact list creation and search functionality.
  | High-quality, low-latency communication with packet or data loss and jitter handling.
  | Support for video calls when sufficient bandwidth is available.
  | Bandwidth adjustment for maintaining communication quality.
  | Provision of secure communication.
  | Provision of reliable signaling, including detection of dropped calls.
  | Notification in case of connection loss.
  | Acoustic echo cancellation (AEC) feature.
  | Voice activity detection (VAD) feature to avoid transmitting voice data when the user is not speaking.
  | Initiation of calls using contact identifiers or contact lists.
  | Display of call status and results (response, in-progress, declined, etc.) during a call.
  | Ability to accept or decline calls when not in an active call. The application should display the caller's contact identifier or name during an incoming call.
  | Ability to end a call at any time during the conversation. Notification should be sent to the other user(s) when one user ends the call.
  | Ability to notify the directory service component of IP changes.
  | Easy addition or replacement of codecs or encryption algorithms by developers/maintainers.
  | Ability to display the application on top during an incoming call.
Server Application: | User registration feature, including password security verification.
  | Functionality for scheduling and supporting video conference calls, with the issuance of conference identifiers to participants.
  | Administrator functionality to verify, deactivate, activate, or delete user accounts.
  | Provision of directory service functionality, mapping contact identifiers to the current IP addresses of devices.
  | Ability to add or replace codecs or encryption algorithms easily by developers/maintainers.
System Software: | Use of Microsoft Windows 10 or 11:
  | Development of client and server applications using languages such as C++, C#, or Python.

QAR
Performance: | The system should provide real-time communication with minimal latency to ensure smooth and uninterrupted voice and video calls.
-- | --
  | The system should efficiently utilize network bandwidth to minimize delays and maximize call quality.
  | The application should handle a reasonable number of concurrent users and maintain performance under typical usage scenarios.
Reliability: | The system should have a high level of reliability, ensuring that users can initiate and maintain communication without frequent disruptions or failures.
  | The system should handle unexpected events, such as network interruptions or server failures, gracefully and recover from them without data loss or significant downtime.
  | The application should have mechanisms to detect and handle dropped calls, reconnecting affected users if necessary.
Security: | The system should provide secure communication, ensuring the confidentiality, integrity, and authenticity of voice and video data transmitted between users.
  | User authentication and authorization mechanisms should be implemented to prevent unauthorized access to the system and protect user accounts.
  | The application should employ encryption algorithms and secure protocols to safeguard sensitive data, such as user credentials and call information.
  | The system should protect against potential security threats, such as eavesdropping, data tampering, or unauthorized interception of calls.
Usability: | The user interface of the application should be intuitive, user-friendly, and visually appealing.
  | Users should be able to easily navigate through the features and functionalities of the application without requiring extensive training or technical knowledge.
  | Error handling should be implemented effectively, providing clear and informative error messages to assist users in troubleshooting issues.
Scalability: | The system should be designed to accommodate future growth and increasing user demand.
  | It should be scalable to handle a larger user base and additional features without compromising performance or quality.
Compatibility: | The application should be compatible with Microsoft Windows 10 and 11 operating systems to ensure broad accessibility for users.
  | Compatibility with standard network protocols and communication technologies should be ensured to enable seamless communication across different devices and networks.

Design constraint
Platform Constraint: | The system should be designed specifically for Microsoft Windows operating systems (Windows 10 and 11).
-- | --
  | The choice of programming languages, frameworks, and libraries should be compatible with the Windows platform.
Technology Constraint: | The project specifies that Session Initiation Protocols (SIP) or any VoIP libraries cannot be used, which limits the available technologies for implementing the voice and video communication system.
  | The project team needs to identify and utilize alternative approaches and technologies for achieving the desired functionality.
Security Constraint: | The system must adhere to security best practices and comply with relevant security standards to ensure the confidentiality, integrity, and authenticity of communication data.
  | Compliance with industry-standard encryption algorithms and secure protocols may be necessary to meet the security requirements.
Performance Constraint: | The system should provide real-time voice and video communication with minimal latency and high quality.
  | The performance should be optimized to handle the expected load and usage patterns without significant degradation in call quality or responsiveness.
Compatibility Constraint: | The system should be compatible with the LG Wi-Fi network infrastructure provided, including the option to connect to individual Wi-Fi routers if necessary.
  | Compatibility with network protocols and configurations specific to LG's infrastructure should be considered during system design.
Resource Constraint: | The project team needs to consider resource limitations, such as memory, processing power, and network bandwidth, when designing the system to ensure optimal utilization of available resources.
Time Constraint: | The project may have specific time constraints, requiring the team to design and implement the system within a given timeframe.
  | The design should consider the project schedule and allocate sufficient time for development, testing, and deployment phases.
Maintainability Constraint: | The system design should prioritize maintainability to facilitate future updates, bug fixes, and enhancements.
  | Code modularity, documentation, and adherence to coding best practices can contribute to the maintainability of the system.

**Comment by ju297** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/user/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/user/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



Type | Type | Description | QA Scenario     (Stimulus / Source of stiulus / Artifact / Environment / Response /   Response measure)
-- | -- | -- | --
QA | Security | System shall handle the user’s password   in secure way | Even if the server data is leaked to the   outside by an insider, it must be encrypted and stored so that the password   is not exposed.     (Data is leaked to the outside / Insider / User's password / All / Store   user's password encrypted / Unknown raw user's password)
QA | Security | System shall transfer data for sessions,   video data, audio streaming data between server & client in secure way. | Even if communication data   between server and client is leaked by an outsider, communication must be   encrypted so that security information is not exposed.     (Data is leaked to the outside / Outsider / Security information / All /   Communication data encryption / Unknown raw security information)
QA | Performance | Provide reliable signaling for both   inbound and outbound calls including dropped call detection. | When a user makes or terminates a direct   call using the system in the runtime environment, the frequency of accurately   responding to the call / end call attempt must be 99.9% or higher.     (Make or Terminate call / User / Call service / Runtime / Respond normally   / More than 99% normal response rate)
QA | Performance | Provide high quality low latency with   strategies for dealing with packet or data loss and jitter. | Jitter is introduced by the gateway, so   the call should operate without a delay of more than one second.     (Jitter occurs / Gateway / Call service / Runtime / It works with no more   than a certain latency / Latency is less than 1 second)
QA | Performance | Adjust bandwidth use as necessary to   maintain communication quality. | Even if the network status of the gateway   deteriorates, video communication is turned off so that voice communication   operates without delay of more than 1 second.     (Deteriorating network conditions / Gateway / Call service / Runtime / Turn   off video communication so that voice communication is maintained / Latency   is less than 1 second)
QA | Modifiability | Provide the ability for the   developer/maintainer to easily add or replace codecs | Changing the codec for video   transmission by customer request should be possible within 5 hours.     (Video codec change request / Customer / Video codec related software   components / During video codec change / System supports the changed video   codec / It takes less than 5 hours to change the video codec)
QA | Modifiability | Provide the ability for the   developer/maintainer to easily add or replace encryption algorithms. | Upon customer request,   encryption algorithm change for data communication must be possible within 5   hours.     (Encryption algorithm change request / Customer / Encryption algorithm   related software components / During encryption algorithm change / System   supports the changed encryption algorithm / It takes less than 5 hours to   change the encryption algorithm)



</body>

</html>

**Reply by ju297** [reply]
QA Scenario 위와 같이 정리했습니다.
크게 변경된 내용은 기존의 "Support a video call if there is enough available bandwidth." 항목을 제거하고 "Adjust bandwidth use as necessary to maintain communication quality." 항목의 QA Scenario로 작성하였습니다.

**Reply by changjurhee** [reply]
wiki에 반영 하도록 할게요.

**Comment by du0san** [comment]
Type | Category | Description
-- | -- | --
FR | User account / Register, update, change | System shall provide to reset user   password in the event it is lost.
FR | User account / Contact lists | System shall provide to create/modify contact list by searching users by   name, first name, address, email or contact identifier

**Reply by du0san** [reply]
첫번째 항목은 FR추가가 필요해보이고 두번째 항목은 create ->  create/modify로 변경하였습니다.
검토 부탁드립니다.
